File: chain/java/CollectingIngredientsChef.java
package chain.java;

public class CollectingIngredientsChef implements Chef {
    private Chef nextChef;

    @Override
    public Chef setNextChef(Chef chef) {
        this.nextChef = chef;
        return chef;
    }

    @Override
    public void cook(String dish) {
        System.out.println("CollectingIngredientsChef is collecting ingredients for: " 
                            + dish);
        if (nextChef != null) {
            nextChef.cook(dish);
        } else {
            System.out.println("No chef available to cook: " + dish);
        }
    }
    
}

File: chain/rust/chain/src/collecting_ingredients_chef.rs
use crate::chef::Chef;
use crate::basic_chef::BasicChef;

pub struct CollectingIngredientsChef {
    pub base: BasicChef,
}

impl Chef for CollectingIngredientsChef {
    fn set_next(&mut self, next: Box<dyn Chef>) { self.base.next = Some(next); }
    fn cook(&self, dish: &str) {
        println!("Collecting ingredients for: {}", dish);
        if let Some(ref next) = self.base.next { next.cook(dish); }
    }
}
